### DISCUSSION TITLE
How do i extract a method ?

### DISCUSSION THREAD

**Discussion by bro69noscope** [neoclide/coc.nvim#5124] [Q&A] [answered]
Been looking for hours I cannot find a way lol

**Comment by fannheyward** [comment]
This needs the language server support the feature first.

**Reply by bro69noscope** [reply]
Yeah I got coc-pyright (Coc-Install done) with rope etc, I just can't figure out the command to use over the visual selection (from either coc in general, or coc-pyright)

``` lua
private extractActions(document: TextDocument, range: Range): CodeAction[] {
    return [
      // extract actions should only work on range text
      {
        title: 'Extract Method',
        kind: CodeActionKind.RefactorExtract,
        command: {
          command: 'python.refactorExtractMethod',
          title: '',
          arguments: [document, range],
        },
      },
```

Like I see that but idk how do I make use of it ? Regular lsp native code action ? Is there some special coc refactor commmand ? Some pyright one ? I don't see them clearly listen in docs anywhere like: "use that vim cmd to call for an extraction". Sorry for the dumb question :sweat:

**Reply by fannheyward** [reply, marked as answer]
```
xmap <leader>a  <Plug>(coc-codeaction-selected)
nmap <leader>a  <Plug>(coc-codeaction-selected)
```

in your vimrc/init.vim first.

`V` to select Python codes, then `<leader>a` to fire code actions, you will get the list, choose the `Extract Method` action:

<img width="210" alt="image" src="https://github.com/user-attachments/assets/40ea9cc1-9b8a-44c9-9a96-c8a18ce7a852">

**Reply by bro69noscope** [reply]
OKAY... I don't know how I didn't see that: thank you very much :heart:
